File: sdks/python/workersdk/worker.py
import json
import time
from typing import Callable, Dict
from kafka import KafkaConsumer, KafkaProducer
from .models import JobMessage, ResultMessage

class Worker:
    def __init__(self):
        self.handlers: Dict[str, Callable[[str], str]] = {}
        self.consumer = None
        self.producer = None

    def register(self, action: str, handler: Callable[[str], str]):
        self.handlers[action] = handler

    def start(self, bootstrap_servers: str = 'kafka:9092'):
        # Retry connection to Kafka
        for attempt in range(30):
            try:
                self.consumer = KafkaConsumer(
                    'workflow-jobs',
                    bootstrap_servers=bootstrap_servers,
                    group_id='workflow-workers',
                    auto_offset_reset='earliest',
                    value_deserializer=lambda m: json.loads(m.decode('utf-8'))
                )
                self.producer = KafkaProducer(
                    bootstrap_servers=bootstrap_servers,
                    value_serializer=lambda v: json.dumps(v).encode('utf-8')
                )
                break
            except Exception as e:
                print(f"Kafka not ready, retrying ({attempt + 1}/30)...")
                time.sleep(2)
        else:
            raise Exception("Could not connect to Kafka after 30 attempts")

        print("Worker started, listening for jobs...")

        for message in self.consumer:
            job = JobMessage.from_dict(message.value)

            print(f"Received job: {job}")

            handler = self.handlers.get(job.action)
            if not handler:
                print(f"No handler for {job.action}")
                continue

            try:
                result = handler(job.payload)
            except Exception as e:
                result = f'{{"error": "{str(e)}"}}'

            response = ResultMessage(job.workflow_run_id, job.action, result)

            self.producer.send('workflow-results', value=response.to_dict())

        self.producer.flush()

File: sdks/java/src/main/java/com/creedpetitt/workersdk/Worker.java
package com.creedpetitt.workersdk;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;

import java.time.Duration;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

public class Worker {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private final Map<String, Handler> handlers = new HashMap<>();
    private KafkaConsumer<String, String> consumer;
    private KafkaProducer<String, String> producer;
    private String bootstrapServers;

    public void register(String action, Handler handler) {
        handlers.put(action, handler);
    }

    public void start() {
        start("kafka:9092");
    }

    public void start(String bootstrapServers) {

        this.bootstrapServers = bootstrapServers;

        // Init consumer and producer
        this.consumer = new KafkaConsumer<>(getConsumerProps());
        this.producer = new KafkaProducer<>(getProducerProps());

        consumer.subscribe(Collections.singleton("workflow-jobs"));

        System.out.println("Workflow started, listening for jobs...");

        while (true) {
            var records = consumer.poll(Duration.ofMillis(100));

            for (ConsumerRecord<String, String> record : records) {
                JobMessage msg = deserializeJob(record.value());

                System.out.println("Received job: " + msg);

                Handler handler = handlers.get(msg.action());
                if (handler == null) {
                    System.err.println("No handler for " + msg.action());
                    continue;
                }

                String output;
                try {
                    output = handler.handle(msg.payload());
                } catch (Exception e) {
                    output = "{\"error\":\"" + e.getMessage() + "\"}";
                }

                ResultMessage res =
                        new ResultMessage(msg.workflowRunId(), msg.action(), output);

                producer.send(
                        new ProducerRecord<>("workflow-results", serializeResult(res))
                );
            }
        }
    }

    // Serialize/deserialize helper methods
    private String serializeResult(ResultMessage result) {
        try {
            return MAPPER.writeValueAsString(result);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private JobMessage deserializeJob(String json) {
        try {
            return MAPPER.readValue(json, JobMessage.class);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    // Kafka consumer/producer properties
    private Properties getConsumerProps() {
        Properties consumerProps = new Properties();
        consumerProps.put("bootstrap.servers", bootstrapServers);
        consumerProps.put("key.deserializer",
                "org.apache.kafka.common.serialization.StringDeserializer");
        consumerProps.put("value.deserializer",
                "org.apache.kafka.common.serialization.StringDeserializer");
        consumerProps.put("group.id", "workflow-workers");
        consumerProps.put("auto.offset.reset", "earliest");

        return consumerProps;
    }

    private Properties getProducerProps() {
        Properties producerProps = new Properties();
        producerProps.put("bootstrap.servers", bootstrapServers);
        producerProps.put("key.serializer",
                "org.apache.kafka.common.serialization.StringSerializer");
        producerProps.put("value.serializer",
                "org.apache.kafka.common.serialization.StringSerializer");

        return producerProps;
    }
}